src/prometheus_handler/mod.rs

```diff
@@ -87,32 +90,40 @@ pub async fn init_prometheus(prometheus_addr: SocketAddr, registry: Registry) ->
 
 /// Init prometheus using the given listener.
 async fn init_prometheus_with_listener(listener: tokio::net::TcpListener, registry: Registry) -> Result<(), Error> {
-    let listener = hyper::server::conn::AddrIncoming::from_listener(listener)?;
-    log::info!("〽️ Prometheus exporter started at {}", listener.local_addr());
+    // let listener = hyper::server::conn::AddrIncoming::from_listener(listener)?;
+    log::info!("〽️ Prometheus exporter started at {}", listener.local_addr().unwrap());
 
-    let service = make_service_fn(move |_| {
-        let registry = registry.clone();
+    loop {
+        // getting the tcp stream and ignoring the remote address
+        let (tcp, _) = listener.accept().await?;
 
-        async move { Ok::<_, hyper::Error>(service_fn(move |req: Request<Body>| request_metrics(req, registry.clone()))) }
-    });
+        // Use an adapter to access something implementing `tokio::io` traits as if they implement
+        let io = TokioIo::new(tcp);
 
-    let (signal, on_exit) = tokio::sync::oneshot::channel::<()>();
-    let server = Server::builder(listener).serve(service).with_graceful_shutdown(async {
-        let _ = on_exit.await;
-    });
+        // Manufacturing a connection with the service from request_metrics fn
+        let conn = server::conn::auto::Builder::new(TokioExecutor::new()).serve_connection(
+            io,
+            service_fn(move |req: Request<hyper::body::Incoming>| request_metrics(req, registry.clone())),
+        );
 
-    let result = server.await.map_err(Into::into);
+        // awaiting for the res
+        // TODO: converting the Box<dyn Error + Send + Sync> to Error. having troubles with that
+        let res = conn.await.map_err(|_| Error::Http(hyper::http::Error::from("Some error occured")));
```

Comment:
Stuck at this part converting Box<dyn Error + Send + Sync> to Error

Reply:
For simplicity I would say that you could do something like


```rust
#[derive(Debug, thiserror::Error)]
pub enum Error {
    /// Hyper internal error.
    #[error(transparent)]
    Hyper(#[from] hyper::Error),

    /// Http request error.
    #[error(transparent)]
    Http(#[from] Box<dyn std::error::Error + Send + Sync>),

    /// i/o error.
    #[error(transparent)]
    Io(#[from] std::io::Error),

    #[error("Prometheus port {0} already in use.")]
    PortInUse(SocketAddr),
}
```

and then here:

```rust
        let res = conn.await.map_err(|e| Error::Http(e.into()))?;
```

as we just want to capture the error in its simplest format here.

Reply:
Yeah, I changed the code accordingly